selfdrive/controls/monitord.py

```diff
@@ -0,0 +1,111 @@
+#!/usr/bin/env python3
+import gc
+from common.realtime import sec_since_boot, set_realtime_priority
+from common.params import Params, put_nonblocking
+import cereal.messaging as messaging
+from selfdrive.controls.lib.drive_helpers import create_event, EventTypes as ET
+from selfdrive.controls.lib.driver_monitor import DriverStatus, MAX_TERMINAL_ALERTS, MAX_TERMINAL_DURATION
+from selfdrive.locationd.calibration_helpers import Calibration
+from selfdrive.controls.lib.gps_helpers import is_rhd_region
+
+def monitord_thread(sm=None, pm=None):
+  gc.disable()
+
+  # start the loop
+  set_realtime_priority(3)
+
+  params = Params()
+
+  # Pub/Sub Sockets
+  if pm is None:
+    pm = messaging.PubMaster(['monitorState'])
+
+  if sm is None:
+    sm = messaging.SubMaster(['driverMonitoring', 'liveCalibration', 'carState', 'model', 'gpsLocation'], ignore_alive=['gpsLocation'])
+
+  driver_status = DriverStatus()
+  is_rhd = params.get("IsRHD")
+  if is_rhd is not None:
+    driver_status.is_rhd = bool(int(is_rhd))
+
+  sm['liveCalibration'].calStatus = Calibration.INVALID
+  v_cruise_last = 0
+
+  # wait until carState live
+  CS_updated = False
+  while(not CS_updated):
+    sm.update(0)
+    CS_updated = sm.updated['carState']
+
+  # 10Hz <- monitoringd
+  while(True):
+    start_time = sec_since_boot()
+    sm.update(0)
+
+    # Get data from monitoringd
+    if not sm.updated['driverMonitoring']:
+      continue
+
+    CS = sm['carState']
+    events = []
+
+    # GPS coords RHD parsing, once every restart
+    if not driver_status.is_rhd_region_checked and sm.updated['gpsLocation']:
+      is_rhd = is_rhd_region(sm['gpsLocation'].latitude, sm['gpsLocation'].longitude)
+      driver_status.is_rhd_region = is_rhd
+      driver_status.is_rhd_region_checked = True
+      put_nonblocking("IsRHD", "1" if is_rhd else "0")
+
+    # Handle calibration
+    cal_status = sm['liveCalibration'].calStatus
+    cal_rpy = [0,0,0]
+    if cal_status == Calibration.CALIBRATED:
+      rpy = sm['liveCalibration'].rpyCalib
+      if len(rpy) == 3:
+        cal_rpy = rpy
+
+    driver_status.get_pose(sm['driverMonitoring'], cal_rpy, CS.vEgo, CS.cruiseState.enabled)
+
+    # Get model meta
+    if sm.updated['model']:
```

Comment:
Because you continue the loop if `driverMonitoring` is not updated, this will only be true if they were updated at the same time. I would structure the code like:

```python
if sm.updated['liveCalibration']:
  # update calibration
if sm.updated['carState']:
  # update v_cruise + driver_engaged
if sm.updated['gpsLocation'] and not driver_status.is_rhd_region_checked:
 # gps stuff 
if sm.updated['driverMonitoring']:
  # driver monitoring stuff
  # send out packet